File: trady-backend/src/chart/domain/input.intraday.ts
import { TimeSeriesType } from 'chart/model';
import { LoadChartInputBase } from './input.base';

export class LoadChartInputIntraday implements LoadChartInputBase {
  timeSeriesType: Readonly<TimeSeriesType> = 'TIME_SERIES_INTRADAY';

  symbol: Readonly<string>;

  interval: Readonly<string>;

  constructor(symbol: string, interval: string) {
    this.symbol = symbol;
    this.interval = interval;
  }

  toDescriptor(): Readonly<string> {
    return `${this.symbol}.${this.timeSeriesType}.${this.interval}`;
  }

  toReferenceUrl(
    prefix: Readonly<String>,
    apiKey: Readonly<string>,
  ): Readonly<string> {
    return (
      `${prefix}` +
      `?function=${this.timeSeriesType}` +
      `&symbol=${this.symbol}` +
      `&interval=${this.interval}` +
      `&apikey=${apiKey}`
    );
  }

  toTimeSeriesKey(): Readonly<string> {
    return `Time Series (${this.interval})`;
  }
}

File: stock-tracker/src/chart_input/input_intraday.rs
use crate::chart_input::LoadChartInputTrait;

pub struct InputIntraday {
  timeseries_type: &'static str,
  symbol: String,
  interval: String,
}

impl InputIntraday {
  pub fn new(symbol: String, interval: String) -> Self {
    Self {
      timeseries_type: "TIME_SERIES_INTRADAY",
      symbol,
      interval,
    }
  }
}

impl LoadChartInputTrait for InputIntraday {
  fn to_descriptor(&self) -> String {
    format!("{}.{}.{}", self.symbol, self.timeseries_type, self.interval)
  }

  fn to_query_string(&self, prefix: &str, apikey: &str) -> String {
    format!(
      "{}?function={}&symbol={}&interval={}&apikey={}",
      prefix, self.timeseries_type, self.symbol, self.interval, apikey
    )
  }

  fn to_timeseries_key(&self) -> String {
    format!("Time Series ({})", self.interval)
  }

  fn to_lifetime_as_seconds(&self) -> usize {
    60 * 1
  }
}
